src/R2N.jl

```diff
@@ -0,0 +1,277 @@
+  ϵ_subsolver = copy(ϵ_subsolver_init)
+  ϵr = options.ϵr
+  verbose = options.verbose
+  maxIter = options.maxIter
+  maxTime = options.maxTime
+  η1 = options.η1
+  η2 = options.η2
+  γ = options.γ
+  θ = options.θ
+  σmin = options.σmin
+
+  local l_bound, u_bound
+  if has_bounds(f)
+    l_bound = f.meta.lvar
+    u_bound = f.meta.uvar
+  end
+
+  if verbose == 0
+    ptf = Inf
+  elseif verbose == 1
+    ptf = round(maxIter / 10)
+  elseif verbose == 2
+    ptf = round(maxIter / 100)
+  else
+    ptf = 1
+  end
+
+  # initialize parameters
+  σk = max(1 / options.ν, σmin)
+  #σk = σmin
+  xk = copy(x0)
+  hk = h(xk[selected])
+  if hk == Inf
+    verbose > 0 && @info "R2N: finding initial guess where nonsmooth term is finite"
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2N: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ = has_bounds(f) ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+
+  Fobj_hist = zeros(maxIter)
+  Hobj_hist = zeros(maxIter)
+  Complex_hist = zeros(Int, maxIter)
+  if verbose > 0
+    #! format: off
+    @info @sprintf "%6s %8s %8s %8s %7s %7s %8s %7s %7s %7s %7s %1s" "outer" "inner" "f(x)" "h(x)" "√ξ1" "√ξ" "ρ" "σ" "‖x‖" "‖s‖" "‖Bₖ‖" "R2N"
+    #! format: on
+  end
+
+  # main algorithm initialization
+
+  local ξ1
+  k = 0
+
+  fk = obj(f, xk)
+  ∇fk = grad(f, xk)
+  ∇fk⁻ = copy(∇fk)
+
+  quasiNewtTest = isa(f, QuasiNewtonModel)
+  Bk = hess_op(f, xk)
+
+  λmax = opnorm(Bk)
+  νInv = (1 + θ) * (λmax + σk)
+
+  optimal = false
+  tired = k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k = k + 1
+    elapsed_time = time() - start_time
+    Fobj_hist[k] = fk
+    Hobj_hist[k] = hk
+
+    # model for first prox-gradient step and ξ1
+    φ1(d) = ∇fk' * d
+    mk1(d) = φ1(d) + ψ(d)
+
+    # model for subsequent prox-gradient steps and ξ
+    φ(d) = (d' * (Bk * d)) / 2 + ∇fk' * d + σk * dot(d, d) / 2
+
+    ∇φ!(g, d) = begin
+      mul!(g, Bk, d)
+      g .+= ∇fk
+      g .+= σk * d
+      g
+    end
+
+    mk(d) = φ(d) + ψ(d)
+
+    # take first proximal gradient step s1 and see if current xk is nearly stationary
+    # s1 minimizes φ1(s) + ‖s‖² / 2 / ν + ψ(s) ⟺ s1 ∈ prox{νψ}(-ν∇φ1(0)).
+
+    subsolver_options.ν = 1 / νInv
+    prox!(s, ψ, -subsolver_options.ν * ∇fk, subsolver_options.ν)
+    ξ1 = hk - mk1(s) + max(1, abs(hk)) * 10 * eps()
+    ξ1 > 0 || error("R2N: first prox-gradient step should produce a decrease but ξ1 = $(ξ1)")
+
+    if ξ1 ≥ 0 && k == 1
+      ϵ_increment = ϵr * sqrt(ξ1)
+      ϵ += ϵ_increment  # make stopping test absolute and relative
+      ϵ_subsolver += ϵ_increment
+    end
+
+    if sqrt(ξ1) < ϵ
+      # the current xk is approximately first-order stationary
+      optimal = true
+      continue
+    end
+
+    if sub_tol
```

Comment:
Thanks you for the PR! Can you explain what is `sub_tol` in the documentation of your function?

Reply:
Ok